### PR TITLE
Add AS hooks for caPutLog module

### PR DESCRIPTION
**Pull request by MaticPogacnik**
Add write hooks for the single and group pvputs. 

This is needed by caPutLog module, to be able log pvputs as well.

### CODE REVIEW

File: pdbApp/pdbgroup.cpp [hunk 1]
```diff
@@ -349,12 +350,24 @@ void PDBGroupPut::put(pvd::PVStructure::shared_pointer const & value,
     // assume value may be a different struct each time... lot of wasted prep work
     const size_t npvs = channel->pv->members.size();
     std::vector<std::tr1::shared_ptr<PVIF> > putpvif(npvs);
+    std::vector<void *> asWritePvt;
 
     for(size_t i=0; i<npvs; i++)
     {
         PDBGroupPV::Info& info = channel->pv->members[i];
-        if(!info.allowProc) continue;
```
Comment (mdavidsaver) [inline pdbApp/pdbgroup.cpp:356]: Is this removal intentional?

File: pdbApp/pdbgroup.cpp [hunk 2]
```diff
@@ -349,12 +350,24 @@ void PDBGroupPut::put(pvd::PVStructure::shared_pointer const & value,
     // assume value may be a different struct each time... lot of wasted prep work
     const size_t npvs = channel->pv->members.size();
     std::vector<std::tr1::shared_ptr<PVIF> > putpvif(npvs);
+    std::vector<void *> asWritePvt;
```
Comment (mdavidsaver) [inline pdbApp/pdbgroup.cpp:353]: Where possible, I try to write exception safe code.  This is relevant here as calls like `ret |= putpvif[i]->get(*changed, doProc);` below can and do throw exceptions.

One solution to avoid this would be to define a new type to own asWrite pointers.  eg.

```c++
struct AsWritePvt {
    void *pvt;
    explicit AsWritePvt(void* pvt): pvt(pvt) {}
    ~AsWritePvt() {
        asTrapWriteAfterWrite(pvt);
    }
};
```

And then define `std::vector<AsWritePvt>`.

Comment (MaticPogacnik) [reply pdbApp/pdbgroup.cpp:353]: Good point,

Where do you prefer this new structure to be defined? It should be in one of the headers that's included in both `pdbsignle.cpp` and `pdbgroup.cpp` as I would use the same approach on both places. Maybe `pdb.h`?

File: pdbApp/pdbgroup.cpp (hunk 1 above)
Comment (MaticPogacnik) [reply pdbApp/pdbgroup.cpp:356]: Hi,
thank you for review.

This was not removed, only GitHub is marking it badly, it still there bellow added code: https://github.com/epics-base/pva2pva/pull/37/files#diff-bd4da3f325a325509ce78293ec376f10R370

I was not sure if trap should come before or after this call. This way notification is sent also if user is not allowed to write if I understand correctly.

File: pdbApp/pdb.h [hunk 3]
```diff
@@ -70,4 +71,13 @@ struct QSRV_API PDBProvider : public epics::pvAccess::ChannelProvider,
 QSRV_API
 void QSRVRegistrar_counters();
 
+class QSRV_API AsWritePvt {
```
Comment (mdavidsaver) [inline pdbApp/pdb.h:74]: A class only needs `QSRV_API` when some methods are defined in .cpp files.

File: pdbApp/pdbgroup.cpp [hunk 4]
```diff
@@ -349,12 +350,24 @@ void PDBGroupPut::put(pvd::PVStructure::shared_pointer const & value,
...
     const size_t npvs = channel->pv->members.size();
     std::vector<std::tr1::shared_ptr<PVIF> > putpvif(npvs);
+    std::vector<AsWritePvt> asWritePvt;
 
     for(size_t i=0; i<npvs; i++)
     {
         PDBGroupPV::Info& info = channel->pv->members[i];
-        if(!info.allowProc) continue;
 
+        asWritePvt.push_back(AsWritePvt(
```
Comment (mdavidsaver) [inline pdbApp/pdbgroup.cpp:359]: Ah, I hadn't thought through all of the implications when providing a prospective definition for `AsWritePvt`.  Specifically, that definition can't be safely copied.  Using `push_back()` is making a copy, and so strictly speaking (with c++98) the destructor should run twice, and result in a double free error.

The fact that it does work in many cases is due to compilers (GCC and recent MSVC) eliding the copy.  However, this is not required.

Since I'm not (yet) requiring c++11 support, it isn't possible to use a move constructor.  An older way to handle this is with `swap()` as can be seen with `class DBCH` in pvif.h (which wraps a `dbChannel*`).

```c++
class AsWritePvt {
    void * pvt;
public:
    AsWritePvt() :pvt(0) {}
    explicit AsWritePvt(void * pvt): pvt(pvt) {}
    ~AsWritePvt() {
        if(pvt) asTrapWriteAfterWrite(pvt);
    }
    void swap(AsWritePvt& o) {
        std::swap(pvt, o.pvt);
    }
private:
    AsWritePvt(const AsWritePvt&);
    AsWritePvt& operator=(const AsWritePvt&);
};
```

Usage is then a bit awkward.

```c++
AsWritePvt temp(...);
asWritePvt.push_back(AsWritePvt());
asWritePvt.back().swap(temp);
```

**Review by mdavidsaver — CHANGES_REQUESTED** [review]

File: pdbApp/pdbgroup.cpp (hunk 4 above)
Comment (MaticPogacnik) [reply pdbApp/pdbgroup.cpp:359]: I had to make the copy constructor public otherwise `asWritePvt.push_back(AsWritePvt());` fails as it cannot make a copy (of an empty object). I am setting the pointer to `NULL` with copy constructor (and I made it explicit). 

Does this look good to you?
```
class AsWritePvt {
    void * pvt;
public:
    AsWritePvt() :pvt(0) {}
    explicit AsWritePvt(const AsWritePvt&) : pvt(NULL) {}
    explicit AsWritePvt(void * pvt): pvt(pvt) {}
    ~AsWritePvt() {
        if(pvt) asTrapWriteAfterWrite(pvt);
    }
    void swap(AsWritePvt& o) {
        std::swap(pvt, o.pvt);
    }
private:
    AsWritePvt& operator=(const AsWritePvt&);
};
```

Thanks

**Comment by mdavidsaver** [comment]
Applied as 69a7bb37956031c32c3deb91d44b80490122b34d with additional fixups cb1b41c1b23ed63df2ac4afca87a7e9de6ddf423

File: pdbApp/pdbgroup.cpp (hunk 4 above)
Comment (mdavidsaver) [reply pdbApp/pdbgroup.cpp:359]: > I had to make the copy constructor public otherwise asWritePvt.push_back(AsWritePvt()); fails

Right, which is why `std::vector` can't be used to store non-copyable types until `std::move()` is available.  Part of the solution is to use `shared_vector`, which has an additional constructor which will default construct each element.  This avoids the need for copying.

>  ... I am setting the pointer to NULL with copy constructor (and I made it explicit).

Various `std::` container classes, including `std::vector` make temporary copies during some operations.  Having a copy constructor which doesn't copy will cause these to break is non-obvious ways.  cf. discussions of the chaos caused by attempts to use `std::vector<std::auto_ptr>`.
